statsmodels/tools/tools.py

```diff
@@ -249,13 +258,14 @@ def _series_add_constant(data, prepend, has_constant):
         columns = [data.name, 'const']
     else:
         columns = ['const', data.name]
-    results = DataFrame({data.name : data, 'const' : const}, columns=columns)
+    results = DataFrame({data.name: data, 'const': const}, columns=columns)
     return results
 
 
 def _dataframe_add_constant(data, prepend, has_constant):
     # check for const.
-    if np.any(data.var(0) == 0):
+
+    if _data_has_constant(data.as_matrix()):
```

Comment:
using `as_matrix()` will upcast and make a copy if needed, which sounds to fragile. (according to the pandas docs/docstring)

For the standard use with `exog` assuming float would be fine, but if users wants to add a constant column to the entire dataframe this might/will break if there are strings or other `object` dtypes.